File: alejo/interaction/gesture_arpeggiator/static/js/hand-tracking.js
/**
 * ALEJO Gesture Arpeggiator - Hand Tracking Client
 * 
 * Handles client-side hand tracking using MediaPipe Hands,
 * with gesture recognition and hand position tracking.
 */

class HandTracking {
    /**
     * Initialize the hand tracking system
     * @param {string} videoElementId - ID of the video element
     * @param {string} canvasElementId - ID of the canvas element for drawing
     * @param {object} options - Configuration options
     */
    constructor(videoElementId, canvasElementId, options = {}) {
        // Elements
        this.videoElement = document.getElementById(videoElementId);
        this.canvasElement = document.getElementById(canvasElementId);
        this.canvasCtx = this.canvasElement.getContext('2d');
        
        // MediaPipe objects
        this.hands = null;
        this.camera = null;
        
        // Tracking state
        this.isRunning = false;
        this.lastHandData = [];
        this.gestureHistory = [];
        this.historyMaxLength = 10;
        
        // Configuration
        this.options = {
            maxHands: options.maxHands || 2,
            minDetectionConfidence: options.minDetectionConfidence || 0.7,
            minTrackingConfidence: options.minTrackingConfidence || 0.5,
            width: 640,
            height: 480,
            fps: 30
        };
        
        // Gesture recognition thresholds
        this.gestureThresholds = {
            pinchThreshold: 0.05,
            extendedFingerThreshold: -0.1,
            palmFacingThreshold: 0.2
        };
        
        // Initialize canvas dimensions
        this.resizeCanvas();
        
        // Bind methods
        this.onResults = this.onResults.bind(this);
        
        // Set up resize listener
        window.addEventListener('resize', () => this.resizeCanvas());
    }
    
    /**
     * Start hand tracking
     * @returns {Promise} Resolves when started
     */
    async start() {
        if (this.isRunning) return;
        
        try {
            console.log('Starting hand tracking...');
            
            // Check if MediaPipe is available
            if (!window.Hands) {
                throw new Error('MediaPipe Hands not available');
            }
            
            // Initialize MediaPipe Hands
            this.hands = new Hands({
                locateFile: (file) => {
                    return `https://cdn.jsdelivr.net/npm/@mediapipe/hands/${file}`;
                }
            });
            
            // Configure MediaPipe Hands
            await this.hands.setOptions({
                maxNumHands: this.options.maxHands,
                modelComplexity: 1,
                minDetectionConfidence: this.options.minDetectionConfidence,
                minTrackingConfidence: this.options.minTrackingConfidence
            });
            
            // Set up result handler
            this.hands.onResults(this.onResults);
            
            // Initialize camera
            this.camera = new Camera(this.videoElement, {
                onFrame: async () => {
                    if (this.isRunning) {
                        await this.hands.send({image: this.videoElement});
                    }
                },
                width: this.options.width,
                height: this.options.height,
                fps: this.options.fps
            });
            
            // Start camera
            await this.camera.start();
            
            this.isRunning = true;
            console.log('Hand tracking started');
            
            return true;
        } catch (error) {
            console.error('Failed to start hand tracking:', error);
            this.isRunning = false;
            throw error;
        }
    }
    
    /**
     * Stop hand tracking
     */
    stop() {
        if (!this.isRunning) return;
        
        console.log('Stopping hand tracking...');
        
        // Stop camera if it exists
        if (this.camera) {
            this.camera.stop();
        }
        
        // Clear canvas
        this.canvasCtx.clearRect(0, 0, this.canvasElement.width, this.canvasElement.height);
        
        this.isRunning = false;
        this.lastHandData = [];
        
        console.log('Hand tracking stopped');
    }
    
    /**
     * Handle results from MediaPipe Hands
     * @param {object} results - Results from MediaPipe Hands
     */
    onResults(results) {
        if (!this.isRunning) return;
        
        // Clear canvas
        this.canvasCtx.save();
        this.canvasCtx.clearRect(0, 0, this.canvasElement.width, this.canvasElement.height);
        
        // Draw hand landmarks
        if (results.multiHandLandmarks && results.multiHandedness) {
            // Process each detected hand
            const handData = [];
            
            for (let i = 0; i < results.multiHandLandmarks.length; i++) {
                const landmarks = results.multiHandLandmarks[i];
                const handedness = results.multiHandedness[i];
                
                // Draw landmarks
                this.drawLandmarks(landmarks, handedness);
                
                // Process hand data
                const processedHand = this.processHandData(landmarks, handedness, i);
                handData.push(processedHand);
                
                // Update gesture indicator
                const elementId = processedHand.isLeft ? 'leftHandGesture' : 'rightHandGesture';
                const element = document.getElementById(elementId);
                if (element) {
                    element.textContent = `${processedHand.isLeft ? 'Left' : 'Right'}: ${processedHand.gesture}`;
                }
            }
            
            // Update last hand data
            this.lastHandData = handData;
            
            // Update gesture history
            if (handData.length > 0) {
                this.gestureHistory.push(handData.map(hand => hand.gesture));
                if (this.gestureHistory.length > this.historyMaxLength) {
                    this.gestureHistory.shift();
                }
            }
        } else {
            // No hands detected
            this.lastHandData = [];
            
            // Update gesture indicators
            document.getElementById('leftHandGesture').textContent = 'Left: None';
            document.getElementById('rightHandGesture').textContent = 'Right: None';
        }
        
        this.canvasCtx.restore();
    }
    
    /**
     * Process hand data from landmarks
     * @param {Array} landmarks - Hand landmarks
     * @param {object} handedness - Hand handedness
     * @param {number} index - Hand index
     * @returns {object} Processed hand data
     */
    processHandData(landmarks, handedness, index) {
        // Determine if left or right hand
        const isLeft = handedness.label === 'Left';
        
        // Extract key landmarks
        const wrist = landmarks[0];
        const thumbTip = landmarks[4];
        const indexTip = landmarks[8];
        const middleTip = landmarks[12];
        const ringTip = landmarks[16];
        const pinkyTip = landmarks[20];
        
        // Calculate hand center
        const centerX = landmarks.reduce((sum, lm) => sum + lm.x, 0) / landmarks.length;
        const centerY = landmarks.reduce((sum, lm) => sum + lm.y, 0) / landmarks.length;
        const centerZ = landmarks.reduce((sum, lm) => sum + lm.z, 0) / landmarks.length;
        
        // Recognize gesture
        const gesture = this.recognizeGesture(landmarks);
        
        // Create hand data object
        return {
            id: index,
            isLeft: isLeft,
            position: { x: centerX, y: centerY, z: centerZ },
            landmarks: landmarks.map(lm => ({ x: lm.x, y: lm.y, z: lm.z })),
            gesture: gesture,
            confidence: handedness.score
        };
    }
    
    /**
     * Recognize hand gesture from landmarks
     * @param {Array} landmarks - Hand landmarks
     * @returns {string} Recognized gesture
     */
    recognizeGesture(landmarks) {
        // Get extended fingers
        const extendedFingers = this.getExtendedFingers(landmarks);
        
        // Check for specific gestures
        if (this.isPinching(landmarks)) {
            return 'pinch';
        }
        
        // Check based on extended fingers
        const fingerSum = extendedFingers.reduce((sum, val) => sum + val, 0);
        
        if (fingerSum === 0) {
            return 'fist';
        } else if (fingerSum === 5) {
            // Open hand - check if palm is facing up or down
            return this.isPalmFacingDown(landmarks) ? 'palm_down' : 'open';
        } else if (extendedFingers[1] === 1 && fingerSum === 1) {
            return 'pointing';
        } else if (extendedFingers[1] === 1 && extendedFingers[2] === 1 && fingerSum === 2) {
            return 'victory';
        } else if (extendedFingers[0] === 1 && fingerSum === 1) {
            return 'thumbs_up';
        } else if (extendedFingers[0] === 1 && extendedFingers[4] === 1 && fingerSum === 2) {
            return 'rock';
        }
        
        // Default
        return 'unknown';
    }
    
    /**
     * Check if fingers are extended
     * @param {Array} landmarks - Hand landmarks
     * @returns {Array} Array of 5 booleans indicating extended fingers
     */
    getExtendedFingers(landmarks) {
        // Wrist position
        const wrist = landmarks[0];
        
        // Finger MCP joints (knuckles)
        const mcpJoints = [landmarks[1], landmarks[5], landmarks[9], landmarks[13], landmarks[17]];
        
        // Finger PIP joints (middle joints)
        const pipJoints = [landmarks[2], landmarks[6], landmarks[10], landmarks[14], landmarks[18]];
        
        // Finger tips
        const tips = [landmarks[4], landmarks[8], landmarks[12], landmarks[16], landmarks[20]];
        
        // Check each finger
        return [
            this.isThumbExtended(landmarks[1], landmarks[2], landmarks[3], landmarks[4]),
            this.isFingerExtended(wrist, mcpJoints[1], pipJoints[1], tips[1]),
            this.isFingerExtended(wrist, mcpJoints[2], pipJoints[2], tips[2]),
            this.isFingerExtended(wrist, mcpJoints[3], pipJoints[3], tips[3]),
            this.isFingerExtended(wrist, mcpJoints[4], pipJoints[4], tips[4])
        ];
    }
    
    /**
     * Check if thumb is extended
     * @param {object} cmc - Carpometacarpal joint
     * @param {object} mcp - Metacarpophalangeal joint
     * @param {object} ip - Interphalangeal joint
     * @param {object} tip - Tip of thumb
     * @returns {number} 1 if extended, 0 if not
     */
    isThumbExtended(cmc, mcp, ip, tip) {
        // Vector from mcp to tip
        const vecX = tip.x - mcp.x;
        const vecY = tip.y - mcp.y;
        
        // Vector from mcp to wrist
        const baseVecX = cmc.x - mcp.x;
        const baseVecY = cmc.y - mcp.y;
        
        // Dot product
        const dot = vecX * baseVecX + vecY * baseVecY;
        
        // If dot product is negative, vectors point in opposite directions
        return dot < 0 ? 1 : 0;
    }
    
    /**
     * Check if finger is extended
     * @param {object} wrist - Wrist landmark
     * @param {object} mcp - Metacarpophalangeal joint
     * @param {object} pip - Proximal interphalangeal joint
     * @param {object} tip - Finger tip
     * @returns {number} 1 if extended, 0 if not
     */
    isFingerExtended(wrist, mcp, pip, tip) {
        // Check if tip is higher than PIP joint (y is inverted in screen coordinates)
        return tip.y < pip.y ? 1 : 0;
    }
    
    /**
     * Check if hand is making a pinch gesture
     * @param {Array} landmarks - Hand landmarks
     * @returns {boolean} True if pinching
     */
    isPinching(landmarks) {
        const thumbTip = landmarks[4];
        const indexTip = landmarks[8];
        
        // Calculate distance between thumb and index fingertips
        const distance = Math.sqrt(
            Math.pow(thumbTip.x - indexTip.x, 2) +
            Math.pow(thumbTip.y - indexTip.y, 2) +
            Math.pow(thumbTip.z - indexTip.z, 2)
        );
        
        return distance < this.gestureThresholds.pinchThreshold;
    }
    
    /**
     * Check if palm is facing down
     * @param {Array} landmarks - Hand landmarks
     * @returns {boolean} True if palm is facing down
     */
    isPalmFacingDown(landmarks) {
        // Use wrist and MCP joints to determine palm orientation
        const wrist = landmarks[0];
        const indexMcp = landmarks[5];
        const pinkyMcp = landmarks[17];
        
        // Calculate normal vector to palm using cross product
        const v1x = indexMcp.x - wrist.x;
        const v1y = indexMcp.y - wrist.y;
        const v1z = indexMcp.z - wrist.z;
        
        const v2x = pinkyMcp.x - wrist.x;
        const v2y = pinkyMcp.y - wrist.y;
        const v2z = pinkyMcp.z - wrist.z;
        
        // Cross product
        const normalX = v1y * v2z - v1z * v2y;
        const normalY = v1z * v2x - v1x * v2z;
        const normalZ = v1x * v2y - v1y * v2x;
        
        // If z component of normal vector is negative, palm is facing down
        return normalZ < this.gestureThresholds.palmFacingThreshold;
    }
    
    /**
     * Draw hand landmarks on canvas
     * @param {Array} landmarks - Hand landmarks
     * @param {object} handedness - Hand handedness
     */
    drawLandmarks(landmarks, handedness) {
        // Draw connections
        this.canvasCtx.lineWidth = 2;
        this.canvasCtx.strokeStyle = handedness.label === 'Left' ? '#00FF00' : '#FF0000';
        
        // Draw connections between landmarks
        this.drawConnections(landmarks);
        
        // Draw landmarks
        this.canvasCtx.fillStyle = handedness.label === 'Left' ? '#00FF00' : '#FF0000';
        
        for (const landmark of landmarks) {
            this.canvasCtx.beginPath();
            this.canvasCtx.arc(
                landmark.x * this.canvasElement.width,
                landmark.y * this.canvasElement.height,
                3,
                0,
                2 * Math.PI
            );
            this.canvasCtx.fill();
        }
    }
    
    /**
     * Draw connections between landmarks
     * @param {Array} landmarks - Hand landmarks
     */
    drawConnections(landmarks) {
        // Define connections
        const connections = [
            // Thumb
            [0, 1], [1, 2], [2, 3], [3, 4],
            // Index finger
            [0, 5], [5, 6], [6, 7], [7, 8],
            // Middle finger
            [0, 9], [9, 10], [10, 11], [11, 12],
            // Ring finger
            [0, 13], [13, 14], [14, 15], [15, 16],
            // Pinky
            [0, 17], [17, 18], [18, 19], [19, 20],
            // Palm
            [0, 5], [5, 9], [9, 13], [13, 17]
        ];
        
        // Draw each connection
        for (const connection of connections) {
            const [i, j] = connection;
            
            this.canvasCtx.beginPath();
            this.canvasCtx.moveTo(
                landmarks[i].x * this.canvasElement.width,
                landmarks[i].y * this.canvasElement.height
            );
            this.canvasCtx.lineTo(
                landmarks[j].x * this.canvasElement.width,
                landmarks[j].y * this.canvasElement.height
            );
            this.canvasCtx.stroke();
        }
    }
    
    /**
     * Resize canvas to match video dimensions
     */
    resizeCanvas() {
        const container = this.canvasElement.parentElement;
        if (!container) return;
        
        const containerWidth = container.clientWidth;
        const containerHeight = container.clientHeight;
        
        // Maintain aspect ratio
        const aspectRatio = this.options.width / this.options.height;
        let width = containerWidth;
        let height = containerWidth / aspectRatio;
        
        if (height > containerHeight) {
            height = containerHeight;
            width = containerHeight * aspectRatio;
        }
        
        this.canvasElement.width = width;
        this.canvasElement.height = height;
    }
    
    /**
     * Update hand tracking and return current hand data
     * @returns {Array} Current hand data
     */
    update() {
        return this.lastHandData;
    }
    
    /**
     * Get the dominant gesture from recent history
     * @returns {string} Dominant gesture
     */
    getDominantGesture() {
        if (this.gestureHistory.length === 0) {
            return 'unknown';
        }
        
        // Count gesture occurrences
        const gestureCounts = {};
        
        for (const handGestures of this.gestureHistory) {
            for (const gesture of handGestures) {
                if (gesture !== 'unknown') {
                    gestureCounts[gesture] = (gestureCounts[gesture] || 0) + 1;
                }
            }
        }
        
        // Find most frequent gesture
        let dominantGesture = 'unknown';
        let maxCount = 0;
        
        for (const [gesture, count] of Object.entries(gestureCounts)) {
            if (count > maxCount) {
                maxCount = count;
                dominantGesture = gesture;
            }
        }
        
        return dominantGesture;
    }
    
    /**
     * Update tracking options
     * @param {object} options - New options
     */
    updateOptions(options) {
        Object.assign(this.options, options);
        
        // Update MediaPipe options if running
        if (this.isRunning && this.hands) {
            this.hands.setOptions({
                maxNumHands: this.options.maxHands,
                minDetectionConfidence: this.options.minDetectionConfidence,
                minTrackingConfidence: this.options.minTrackingConfidence
            });
        }
    }
}

File: alejo/interaction/gesture_arpeggiator/hand_tracking.py
"""
Hand Tracking Module for Gesture Arpeggiator

This module implements hand tracking and gesture recognition using MediaPipe,
providing real-time hand landmark detection and gesture classification for
ALEJO's gesture-based arpeggiator system.
"""

import logging
import numpy as np
import cv2
import math
from typing import Dict, List, Optional, Tuple, Any, Union
from enum import Enum
import asyncio
import time
import json

# Import MediaPipe dependencies conditionally to allow for graceful fallback
try:
    import mediapipe as mp
    MEDIAPIPE_AVAILABLE = True
except ImportError:
    MEDIAPIPE_AVAILABLE = False
    logging.warning("MediaPipe not available. Hand tracking will be disabled.")

logger = logging.getLogger(__name__)

class GestureType(str, Enum):
    """Recognized hand gestures for musical control"""
    OPEN = "open"
    CLOSED = "closed"
    POINTING = "pointing"
    PINCH = "pinch"
    VICTORY = "victory"
    THUMBS_UP = "thumbs_up"
    ROCK = "rock"
    PALM_DOWN = "palm_down"
    PALM_UP = "palm_up"
    FIST = "fist"
    UNKNOWN = "unknown"

class HandTracker:
    """
    Hand tracking and gesture recognition using MediaPipe
    
    This class provides real-time hand landmark detection and gesture
    classification for the gesture arpeggiator system.
    """
    
    def __init__(self, 
                 max_num_hands: int = 2,
                 min_detection_confidence: float = 0.7,
                 min_tracking_confidence: float = 0.5):
        """
        Initialize the hand tracker
        
        Args:
            max_num_hands: Maximum number of hands to detect
            min_detection_confidence: Minimum confidence for hand detection
            min_tracking_confidence: Minimum confidence for hand tracking
        """
        self.max_num_hands = max_num_hands
        self.min_detection_confidence = min_detection_confidence
        self.min_tracking_confidence = min_tracking_confidence
        
        self._initialized = False
        self._mp_hands = None
        self._hands = None
        self._mp_drawing = None
        self._mp_drawing_styles = None
        
        # Hand tracking state
        self.current_hands: List[Dict] = []
        self.hand_history: List[List[Dict]] = []
        self.history_max_length = 30  # Store 1 second at 30fps
        
        # Initialize MediaPipe if available
        if MEDIAPIPE_AVAILABLE:
            self._initialize_mediapipe()
        
    def _initialize_mediapipe(self):
        """Initialize MediaPipe Hands solution"""
        try:
            mp_hands = mp.solutions.hands
            mp_drawing = mp.solutions.drawing_utils
            mp_drawing_styles = mp.solutions.drawing_styles
            
            # Initialize the Hands solution
            hands = mp_hands.Hands(
                static_image_mode=False,
                max_num_hands=self.max_num_hands,
                min_detection_confidence=self.min_detection_confidence,
                min_tracking_confidence=self.min_tracking_confidence
            )
            
            self._mp_hands = mp_hands
            self._hands = hands
            self._mp_drawing = mp_drawing
            self._mp_drawing_styles = mp_drawing_styles
            self._initialized = True
            
            logger.info("MediaPipe Hands initialized successfully")
        except Exception as e:
            logger.error(f"Failed to initialize MediaPipe Hands: {e}")
            self._initialized = False
            
    def is_initialized(self) -> bool:
        """Check if the hand tracker is initialized"""
        return self._initialized
        
    def process_frame(self, frame: np.ndarray) -> Tuple[List[Dict], np.ndarray]:
        """
        Process a video frame to detect and track hands
        
        Args:
            frame: Input video frame (BGR format)
            
        Returns:
            Tuple of (hand_data, annotated_frame)
        """
        if not self._initialized:
            return [], frame
            
        # Convert BGR to RGB
        frame_rgb = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
        
        # Process the frame with MediaPipe Hands
        results = self._hands.process(frame_rgb)
        
        # Convert back to BGR for OpenCV
        annotated_frame = cv2.cvtColor(frame_rgb, cv2.COLOR_RGB2BGR)
        
        # Extract hand data
        hand_data = self._extract_hand_data(results, frame.shape)
        
        # Update hand history
        self.current_hands = hand_data
        self.hand_history.append(hand_data)
        if len(self.hand_history) > self.history_max_length:
            self.hand_history.pop(0)
            
        # Draw hand landmarks on the frame
        if results.multi_hand_landmarks:
            for hand_landmarks in results.multi_hand_landmarks:
                self._mp_drawing.draw_landmarks(
                    annotated_frame,
                    hand_landmarks,
                    self._mp_hands.HAND_CONNECTIONS,
                    self._mp_drawing_styles.get_default_hand_landmarks_style(),
                    self._mp_drawing_styles.get_default_hand_connections_style()
                )
                
        return hand_data, annotated_frame
        
    def _extract_hand_data(self, results, frame_shape) -> List[Dict]:
        """
        Extract hand data from MediaPipe results
        
        Args:
            results: MediaPipe Hands processing results
            frame_shape: Shape of the input frame
            
        Returns:
            List of hand data dictionaries
        """
        hands = []
        
        if not results.multi_hand_landmarks:
            return hands
            
        height, width, _ = frame_shape
        
        # Process each detected hand
        for idx, (hand_landmarks, handedness) in enumerate(
            zip(results.multi_hand_landmarks, results.multi_handedness)
        ):
            # Extract landmarks
            landmarks = []
            for landmark in hand_landmarks.landmark:
                landmarks.append({
                    "x": landmark.x,
                    "y": landmark.y,
                    "z": landmark.z
                })
                
            # Determine if left or right hand
            is_left = handedness.classification[0].label == "Left"
            
            # Calculate hand center position
            center_x = sum(lm["x"] for lm in landmarks) / len(landmarks)
            center_y = sum(lm["y"] for lm in landmarks) / len(landmarks)
            center_z = sum(lm["z"] for lm in landmarks) / len(landmarks)
            
            # Recognize gesture
            gesture = self._recognize_gesture(landmarks)
            
            # Calculate confidence
            confidence = handedness.classification[0].score
            
            # Create hand data dictionary
            hand_data = {
                "id": idx,
                "landmarks": landmarks,
                "gesture": gesture,
                "position": {
                    "x": center_x,
                    "y": center_y,
                    "z": center_z
                },
                "isLeft": is_left,
                "confidence": confidence
            }
            
            hands.append(hand_data)
            
        return hands
        
    def _recognize_gesture(self, landmarks: List[Dict]) -> str:
        """
        Recognize hand gesture from landmarks
        
        Args:
            landmarks: List of hand landmarks
            
        Returns:
            Recognized gesture type
        """
        if not landmarks or len(landmarks) < 21:
            return GestureType.UNKNOWN
            
        # Extract key landmarks
        wrist = landmarks[0]
        thumb_tip = landmarks[4]
        index_tip = landmarks[8]
        middle_tip = landmarks[12]
        ring_tip = landmarks[16]
        pinky_tip = landmarks[20]
        
        # Calculate finger extensions
        fingers_extended = self._get_extended_fingers(landmarks)
        
        # Recognize gestures based on finger positions
        if sum(fingers_extended) == 0:
            return GestureType.FIST
        elif sum(fingers_extended) == 5:
            # Check palm orientation
            if self._is_palm_facing_down(landmarks):
                return GestureType.PALM_DOWN
            else:
                return GestureType.OPEN
        elif fingers_extended == [0, 1, 0, 0, 0]:
            return GestureType.POINTING
        elif fingers_extended == [0, 1, 1, 0, 0]:
            return GestureType.VICTORY
        elif fingers_extended == [1, 0, 0, 0, 0]:
            return GestureType.THUMBS_UP
        elif fingers_extended == [1, 0, 0, 0, 1]:
            return GestureType.ROCK
        elif self._is_pinching(landmarks):
            return GestureType.PINCH
            
        return GestureType.UNKNOWN
        
    def _get_extended_fingers(self, landmarks: List[Dict]) -> List[int]:
        """
        Determine which fingers are extended
        
        Args:
            landmarks: List of hand landmarks
            
        Returns:
            List of binary values indicating extended fingers
            [thumb, index, middle, ring, pinky]
        """
        # Get key points
        wrist = landmarks[0]
        thumb_tip = landmarks[4]
        thumb_ip = landmarks[3]
        thumb_mcp = landmarks[2]
        
        index_tip = landmarks[8]
        index_pip = landmarks[6]
        
        middle_tip = landmarks[12]
        middle_pip = landmarks[10]
        
        ring_tip = landmarks[16]
        ring_pip = landmarks[14]
        
        pinky_tip = landmarks[20]
        pinky_pip = landmarks[18]
        
        # Check if each finger is extended
        # Thumb is special case
        thumb_extended = self._is_thumb_extended(thumb_tip, thumb_ip, thumb_mcp, wrist)
        
        # For other fingers, compare tip position to PIP joint
        index_extended = index_tip["y"] < index_pip["y"]
        middle_extended = middle_tip["y"] < middle_pip["y"]
        ring_extended = ring_tip["y"] < ring_pip["y"]
        pinky_extended = pinky_tip["y"] < pinky_pip["y"]
        
        return [
            1 if thumb_extended else 0,
            1 if index_extended else 0,
            1 if middle_extended else 0,
            1 if ring_extended else 0,
            1 if pinky_extended else 0
        ]
        
    def _is_thumb_extended(self, tip, ip, mcp, wrist) -> bool:
        """
        Check if thumb is extended
        
        Args:
            tip: Thumb tip landmark
            ip: Thumb IP joint landmark
            mcp: Thumb MCP joint landmark
            wrist: Wrist landmark
            
        Returns:
            True if thumb is extended, False otherwise
        """
        # Calculate vectors
        wrist_to_mcp = [mcp["x"] - wrist["x"], mcp["y"] - wrist["y"]]
        mcp_to_tip = [tip["x"] - mcp["x"], tip["y"] - mcp["y"]]
        
        # Calculate angle between vectors
        dot_product = wrist_to_mcp[0] * mcp_to_tip[0] + wrist_to_mcp[1] * mcp_to_tip[1]
        wrist_to_mcp_mag = math.sqrt(wrist_to_mcp[0]**2 + wrist_to_mcp[1]**2)
        mcp_to_tip_mag = math.sqrt(mcp_to_tip[0]**2 + mcp_to_tip[1]**2)
        
        if wrist_to_mcp_mag * mcp_to_tip_mag == 0:
            return False
            
        cos_angle = dot_product / (wrist_to_mcp_mag * mcp_to_tip_mag)
        cos_angle = max(-1.0, min(1.0, cos_angle))  # Clamp to [-1, 1]
        angle = math.acos(cos_angle) * 180 / math.pi
        
        # Thumb is extended if angle is large enough
        return angle > 30
        
    def _is_pinching(self, landmarks: List[Dict]) -> bool:
        """
        Check if hand is making a pinch gesture
        
        Args:
            landmarks: List of hand landmarks
            
        Returns:
            True if pinching, False otherwise
        """
        thumb_tip = landmarks[4]
        index_tip = landmarks[8]
        
        # Calculate distance between thumb and index fingertips
        dx = thumb_tip["x"] - index_tip["x"]
        dy = thumb_tip["y"] - index_tip["y"]
        dz = thumb_tip["z"] - index_tip["z"]
        
        distance = math.sqrt(dx*dx + dy*dy + dz*dz)
        
        # Pinch threshold
        return distance < 0.05
        
    def _is_palm_facing_down(self, landmarks: List[Dict]) -> bool:
        """
        Check if palm is facing down
        
        Args:
            landmarks: List of hand landmarks
            
        Returns:
            True if palm is facing down, False otherwise
        """
        # Use cross product of vectors to determine palm orientation
        wrist = landmarks[0]
        index_mcp = landmarks[5]
        pinky_mcp = landmarks[17]
        
        # Create vectors
        v1 = [index_mcp["x"] - wrist["x"], 
              index_mcp["y"] - wrist["y"], 
              index_mcp["z"] - wrist["z"]]
              
        v2 = [pinky_mcp["x"] - wrist["x"], 
              pinky_mcp["y"] - wrist["y"], 
              pinky_mcp["z"] - wrist["z"]]
              
        # Cross product
        cross = [
            v1[1] * v2[2] - v1[2] * v2[1],
            v1[2] * v2[0] - v1[0] * v2[2],
            v1[0] * v2[1] - v1[1] * v2[0]
        ]
        
        # If z component of cross product is negative, palm is facing down
        return cross[2] < 0
        
    def get_current_hands(self) -> List[Dict]:
        """Get current hand tracking data"""
        return self.current_hands
        
    def get_hand_history(self) -> List[List[Dict]]:
        """Get hand tracking history"""
        return self.hand_history
        
    def release(self):
        """Release resources"""
        if self._initialized and self._hands:
            self._hands.close()
            self._initialized = False
